File: src/main/java/com/zp4rker/iab/InfinityAndBeyond.java
package com.zp4rker.iab;

import com.mongodb.ConnectionString;
import com.mongodb.MongoClientSettings;
import com.mongodb.MongoException;
import com.mongodb.client.MongoClients;
import com.zp4rker.iab.api.Explorer;
import com.zp4rker.iab.api.FlightLog;
import com.zp4rker.iab.api.Planet;
import com.zp4rker.iab.api.Spaceship;
import com.zp4rker.iab.api.storage.CelestialLocationCodec;
import com.zp4rker.iab.api.storage.PlaneteryLocationCodec;
import com.zp4rker.iab.api.storage.UUIDCodec;
import dev.morphia.Datastore;
import dev.morphia.Morphia;
import org.bson.codecs.configuration.CodecRegistries;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.logging.Logger;

public class InfinityAndBeyond extends JavaPlugin {
    public static Logger LOGGER;

    @Override
    public void onEnable() {
        LOGGER = getLogger();

        saveDefaultConfig();

        registerCommands();
        registerListeners();

        setupDatabase();
    }

    private void registerCommands() {
        // Register commands
    }

    private void registerListeners() {
        // Arrays.asList().forEach(listener -> getServer().getPluginManager().registerEvents(listener, this));
    }

    public static Datastore DATABASE;

    private void setupDatabase() {
        String connectionString = getConfig().getString("database.connectionString");
        try {
            assert connectionString != null;
            MongoClientSettings settings = MongoClientSettings.builder().codecRegistry(CodecRegistries.fromRegistries(
                    CodecRegistries.fromCodecs(
                            new CelestialLocationCodec(),
                            new PlaneteryLocationCodec(),
                            new UUIDCodec()
                    ),
                    MongoClientSettings.getDefaultCodecRegistry()
            )).applyConnectionString(new ConnectionString(connectionString)).build();
            DATABASE = Morphia.createDatastore(MongoClients.create(settings), "infinityandbeyond");
            DATABASE.getMapper().map(Explorer.class, Spaceship.class, FlightLog.class, Planet.class);
            DATABASE.ensureIndexes();
        } catch (MongoException | AssertionError e) {
            LOGGER.warning("Invalid connection string, or unable to connect to database server! Shutting down...");
            getServer().getPluginManager().disablePlugin(this);
        }
    }
}

File: src/main/kotlin/com/zp4rker/iab/InfinityAndBeyond.kt
package com.zp4rker.iab

import co.aikar.commands.PaperCommandManager
import com.zp4rker.bukkot.extensions.register
import com.zp4rker.iab.db.DBManager
import com.zp4rker.iab.listeners.WelcomeMessage
import com.zp4rker.iab.prompts.ExplorerSetup
import org.apache.logging.log4j.Level
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.core.Filter
import org.apache.logging.log4j.core.LogEvent
import org.apache.logging.log4j.core.filter.AbstractFilter
import org.bukkit.plugin.java.JavaPlugin
import java.io.File
import java.sql.SQLException

lateinit var IAB: InfinityAndBeyond
var DB_MANAGER: DBManager? = null

class InfinityAndBeyond : JavaPlugin() {
    override fun onEnable() {
        IAB = this

        saveDefaultConfig()

        registerCommands()
        registerListeners()

        configDatabase()
        shushOrmliteLogger()
        createTables()
    }

    override fun onDisable() {
        DB_MANAGER?.closeConnection()
    }

    private fun registerCommands() {
        val manager = PaperCommandManager(this)
        // listOf().forEach { manager.registerCommand(it) }
    }

    private fun registerListeners() {
        arrayOf(
            ExplorerSetup.Listener,
            WelcomeMessage
        ).forEach { it.register(this) }
    }

    private fun configDatabase() {
        var connectionUrl = config.getString("database.connectionUrl", null)
        if (connectionUrl.isNullOrBlank()) {
            val dbType = config.getString("database.type", "mysql")
            val host = config.getString("database.host", null)
            val user = config.getString("database.user", null)
            val password = config.getString("database.password", null)
            val dbname = config.getString("database.dbname", "infinityandbeyond")

            if (host == null || (dbType.equals("mysql", true) && arrayOf(user, password).any { it == null })) {
                logger.warning("Invalid/incomplete database configuration found! Falling back to file database")
                val file = File(dataFolder, "plugindata.db").absolutePath
                connectionUrl = "jdbc:sqlite:$file"
            } else {
                connectionUrl = DBManager.connectionUrl(dbType!!, host, user, password, dbname!!)
            }
        }
        DB_MANAGER = DBManager(connectionUrl)
        logger.info("Completed database configuration")
    }

    private fun createTables() {
        try {
            DB_MANAGER?.let {
                logger.info("Ensuring tables are setup")
                it.createTables()
                logger.info("Completed tables setup")
            }
        } catch (e: SQLException) {
            logger.severe("Unable to create tables: ${e.message}")
            logger.info("Disabling plugin")
            server.pluginManager.disablePlugin(this)
        }
    }

    private fun shushOrmliteLogger() {
        (LogManager.getRootLogger() as org.apache.logging.log4j.core.Logger).addFilter(object : AbstractFilter() {
            override fun filter(event: LogEvent?): Filter.Result {
                event?.let {
                    if (it.loggerName.startsWith("com.j256.ormlite") && it.level.isLessSpecificThan(Level.WARN)) {
                        return Filter.Result.DENY
                    }
                }
                return Filter.Result.NEUTRAL
            }
        })
    }
}